File: lib/backend/src/extract-sa.js
var output_line = '',
    result_ingestion = '',
    //grep = require('simple-grep'),
    fs = require('fs'),
    exec = require('child_process').exec,
    grafana_endpoint='';

var grep = function(what, where, callback){

	exec("grep " + what + " " + where + " -nr", function(err, stdin, stdout){
		var list = {}

		var results = stdin.split('\n');

	    // remove last element (it’s an empty line)
	    results.pop();
	    for (var i = 0; i < results.length; i++) {
	    	var eachPart = results[i].split(':') //file:linenum:line
	    	list[eachPart[0]] = []
	    }
	    for (var i = 0; i < results.length; i++) {
	    	var eachPart = results[i].split(':') //file:linenum:line
	    	var details = {}
	    	var filename = eachPart[0]
	    	details['line_number'] = eachPart[1]

	    	eachPart.shift()
	    	eachPart.shift()
	    	details['line'] = eachPart.join(':')

	    	list[filename].push(details)
	    }


	    var results = []
	    var files = Object.keys(list)
	    for(var i = 0; i < files.length; i++){
	    	results.push({'file' : files[i], 'results' : list[files[i]]})
	    }

	    callback(results)
	});
}

module.exports = {
    extract: function(req, res, _process, util, create, path, zcache, sa_file_path) {
        util.log("sadf type determined: " + zcache["session_data"][req.sessionID]["sadf_type_det"]);
        var commandObj,
            txtCmd = ['LC_ALL=C', 'scripts/vos/analysis/bin/sadf', '-x'];
        txtCmd[1] = txtCmd[1] + '-' + zcache["session_data"][req.sessionID]["sadf_type_det"] + '-' + '64';
        txtCmd.push(sa_file_path);
        txtCmd.push('--');
        var sar_file = path.join(zcache["session_data"][req.sessionID]["datafile_path"], "sar.xml");
        /* by default, extract all. This is because older versions of 'sadf' fail when certain
        orders of args are specified or when the data for arg is not present; the newer
        ones however fail silently and process rest of the args. The workaround includes
        generating all data with -A by default and then choosing the ones selected..
        (with exception handling ofcourse). */
        txtCmd.push('-A');
        util.log(txtCmd.join(' '));
        commandObj = _process.spawn('env', txtCmd);
        util.log("spawned..");

        commandObj.stdout.on('data', function (data) {
            zcache["session_data"][req.sessionID]["stdOut"] += data.toString();
        });
        commandObj.stdout.on('end', function () {

            var nodename_cmd = 'scripts/detect_nodename ' + sar_file;
            var sar_cmd = function() {
                var nodenameObj = _process.spawn('env', nodename_cmd.split(' '));
                nodenameObj.stdout.on('data', function (data) {
                    zcache["session_data"][req.sessionID]["nodename"] += data.toString();
                });
                nodenameObj.stdout.on('end', function(){
                    var ix_res = {};
                    var sarObj, sarCmd = ['scripts/vos/analysis/bin/index-sar',
                                          sar_file,
                                          zcache["session_data"][req.sessionID]["nodename"]];
                    util.log('ES indexing cmd: ' + sarCmd.join(' '));
                    sarObj = _process.spawn('env', sarCmd);
                    sarObj.stdout.on('data', function (data) {
                        output_line = data.toString();
                        match_result = output_line.match(/(grafana_range_begin|grafana_range_end) \d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/g);
                        if(match_result) {
                            for(var i=0; i<match_result.length; i++) {
                                match_result[i] = match_result[i].split(' ');
                                ix_res[match_result[i][0]] = match_result[i][1];
                            }
                        } else {
                            result_ingestion += output_line;
                        }
                    });
                    sarObj.stdout.on('end', function(){
                        util.log("beg: " + ix_res['grafana_range_begin']);
                        util.log("end: " + ix_res['grafana_range_end']);

                        create.dashboard(_process, util, zcache["session_data"][req.sessionID]["nodename"],
                            zcache["session_data"][req.sessionID]["fields"],
                            ix_res['grafana_range_begin'],
                            ix_res['grafana_range_end']);
                    });
                    sarObj.stderr.on('data', function (data) {
                        util.log('stderr: ' + data);
                    });
                    sarObj.stdout.on('close', function(){
                        util.log('[ES data ingested] -- ' + zcache["session_data"][req.sessionID]["nodename"]);
                        if (!!zcache["session_data"][req.sessionID]["nodename"]) {
                            function callback(ff){
                              grep('dashboard_url', '/etc/sar-index.cfg', function(list){
                                var _tmp = list[0].results[0];
                                var line = _tmp.line_number + ":" + _tmp.line;
                                grafana_endpoint = line.split('dashboard_url = ')[1];
                                ff();
                              });
                            };

                            callback(function(){
                              res.write("Success! <u>Nodename: <b>" +
                                        zcache["session_data"][req.sessionID]["nodename"] + "</b></u>");
                              res.write("<p>kindly go to <a href='" + grafana_endpoint +
                                        "'>this Grafana homepage</a> and login with <pre>admin/admin</pre>" +
                                        " to find content for your nodename's result.. <br/>" +
					"(Click on the dropdown menu 'Home' in grafana homepage to find your dashboard) </p>");
                              res.end();
                            });
                        } else {
                            res.write("Fail");
                            res.end();
                        }
                        output_line = '';
                        result_ingestion = '';
                    });
                });
                nodenameObj.stderr.on('data', function (data) {
                    util.log('stderr: ' + data);
                    output_line = '';
                    result_ingestion = '';
                });
            };

            fs.writeFile(sar_file, zcache["session_data"][req.sessionID]["stdOut"], { flags: 'w', encoding: 'utf-8' }, function(err) {
                if(err) {
                    util.log("SAR data extraction *failed*!");
                    return util.log(err);
                } else {
                    util.log("SAR data extraction *completed*!");
                    sar_cmd();
                }
            });

        });
        commandObj.stderr.on('data', function (data) {
            util.log('stderr: ' + data.toString());
            res.write('Sorry : ' + data.toString());
            res.end();
        });

        commandObj.on('close', function (code) {
            zcache["session_data"][req.sessionID]["sadf_type_det"] = '';
            util.log('child process exited with code ' + code);
        });
    }
}

File: lib/backend/src/extract_sa.py
import os
import sys
import requests
import subprocess

# user mods
import creation
from app import app
from scripts.vos.analysis.lib import index_sar


def extract(sessionID, target, sa_filename):
    TSTAMPS={}
    CMD_CONVERT = ['-x', "--", "-A"]

    SAR_XML_FILEPATH = os.path.join(target, "%s.%s" % (sa_filename, "sar.xml"))
    file_metadata = "file_metadata:%s:%s" % (sessionID, sa_filename)

    sadf_type_det = app.cache.hget(file_metadata, "sadf_type_det").decode()
    app.logger.info('sysstat version found for: %s' % sadf_type_det)

    _SCRIPT = "%s-%s-64" % ('scripts/vos/analysis/bin/sadf', sadf_type_det)
    CMD_CONVERT.insert(0, _SCRIPT)

    conv_path = app.cache.hget(file_metadata, "sa_file_path_conv")
    if conv_path:
        target_file = conv_path.decode()
    else:
        target_file = app.cache.hget(file_metadata, "sa_file_path").decode()

    CMD_CONVERT.insert(-2, target_file)

    #FIXME: check if env in Popen is working fine
    app.logger.info("spawned CMD: %s" % " ".join(CMD_CONVERT));
    p1 = subprocess.Popen(CMD_CONVERT, env={'LC_ALL': 'C'},
                            stdout=open(SAR_XML_FILEPATH, 'w'))
    p1.wait()
    app.logger.info("SAR data extraction *completed*!")

    CMD_GREP = ["scripts/detect_nodename", SAR_XML_FILEPATH]
    p2 = subprocess.Popen(CMD_GREP, stdout=subprocess.PIPE)
    p2.wait()
    NODENAME = p2.communicate()[0].decode().replace("\n", "")

    #FIXME: check if call_indexer works everytime. And if it handles errors
    try:
        #FIXME: bad XML ExpatError
        raise
        state, beg, end = index_sar.call_indexer(file_path=SAR_XML_FILEPATH,
                               _nodename=NODENAME,
                               cfg_name=app.config.get('CFG_PATH'),
                               run_unique_id=sessionID,
                               run_md5=sessionID)
        # import pdb; pdb.set_trace()

        if state:
            TSTAMPS['grafana_range_begin'] = beg
            TSTAMPS['grafana_range_end'] = end
    except Exception as E:
        #FIXME: remove if we use the try: approach in future.
        app.logger.warn(E)
        app.logger.warn("=====Running alternate ES indexing script======")
        CMD_INDEXING = ['scripts/vos/analysis/bin/index-sar',
                        SAR_XML_FILEPATH, NODENAME]
        app.logger.info('ES indexing cmd: ' + " ".join(CMD_INDEXING))
        p3 = subprocess.Popen(CMD_INDEXING, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        p3.wait()
        RESULT_RAW = p3.communicate()
        RAW_STDOUT = RESULT_RAW[0].decode().splitlines()
        RAW_STDERR = RESULT_RAW[1].decode()

        if "ConnectionError" in RAW_STDERR:
            print(RAW_STDERR, file=sys.stderr)
            return (NODENAME, False, sadf_type_det)

        if "ERROR" in RAW_STDERR:
            # ES indexing failed
            print(RAW_STDERR, file=sys.stderr)
            # return (NODENAME, False, sadf_type_det)

        TMP = [line for line in RAW_STDOUT if line.startswith('grafana_range')]
        if TMP:
            for line in TMP:
                k,v = line.split(' ')
                TSTAMPS[k] = v

    if TSTAMPS:
        app.logger.info("[ES data ingested] -- %s" % NODENAME);
        app.logger.info('beg: %s' % TSTAMPS['grafana_range_begin'])
        app.logger.info('end: %s' % TSTAMPS['grafana_range_end'])

        GRAPHING_OPTIONS = app.cache.hget("sar_args:%s" % sessionID, "fields").decode()
        creation.dashboard(NODENAME, GRAPHING_OPTIONS, TSTAMPS)

        return (NODENAME, TSTAMPS, sadf_type_det)
    else:
        return (NODENAME, False, sadf_type_det)
